src/dataflow/src/decode/mod.rs

```diff
@@ -292,6 +292,84 @@ where
     (oks.as_collection(), Some(errs.as_collection()))
 }
 
+/// Update row to blank out retractions of rows that we have never seen
+fn rewrite_for_upsert(
+    val: Result<Row, DataflowError>,
+    keys: &mut HashMap<Box<[u8]>, Row>,
+    key: &[u8],
+) -> Result<Row, DataflowError> {
+    if let Ok(row) = val {
+        // TODO: handle parsed keys
+        let entry = keys.entry(key.into());
+
+        let mut rowiter = row.iter();
+        let before = rowiter.next().expect("must have a before list");
+        let after = rowiter.next().expect("must have an after list");
+
+        assert_eq!(
+            rowiter.next(),
+            None,
+            "[customer-data] Debezium data should only have retract/insert lists, got unexpected third: {:?}",
+            row
+        );
+        assert!(
+            matches!(before, Datum::List { .. } | Datum::Null),
+            "[customer-data] Debezium logic should be a List or absent, got {:?}",
+            before
+        );
+
+        match entry {
+            Entry::Vacant(vacant) => {
+                // if the key is new, then we know that we always need to ignore the "before" part,
+                // so zero it out
+                vacant.insert({
+                    let mut packer = RowPacker::new();
+                    packer.push(after);
+                    packer.finish()
+                });
+
+                if before.is_null() {
+                    Ok(row)
+                } else {
+                    let mut packer = RowPacker::new();
+                    packer.push(Datum::Null);
+                    packer.push(after);
+                    Ok(packer.finish())
+                }
+            }
+            Entry::Occupied(mut occupied) => {
+                if occupied.get().iter().next() == Some(before) {
+                    if after.is_null() {
```

Comment:
Don't we need an `else` branch here to set the value in the map to the new value?

Reply:
ah, yes, I'll add a test that checks for that too.